### ISSUE TITLE
Implement back-off logic for nodes reconnecting too quickly

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#5319]
In some scenarios, networking issues can make nodes reconnect and disconnect in quick succession. This can cause tables to become and remain unavailable, if the node connecting triggers a Raft election and/or a primary handover.

The idea for the back-off is as follows:
On each node, we maintain two maps:
```C++
std::map<server_id_t, ticks_t> reconnect_delay;
std::map<server_id_t, ticks_t> last_connect;
```

When a node connects, we follow the following logic:
```
if (last_connect[server] + reconnect_delay[server] < current_ticks) {
    reject_connection();
}
last_connect[server] = current_ticks;
reconnect_delay[server] = default_delay; // default_delay will be a small number, maybe 100ms or something
```

When a node disconnects, we follow this logic:
```
if (last_connect[server] + reconnect_delay[server] < current_ticks) {
    // If the server keeps dropping off too quickly, increase the delay.
    // The factor of 2 is arbitrary.
    reconnect_delay[server] = std::min(max_delay, reconnect_delay[server] * 2);
}
```

**Comment by danielmewes** [comment]
(we noted the exact logic above doesn't work (especially the reset of `reconnect_delay` after connecting). So ignore the details of it for now)

**Comment by VeXocide** [comment]
In CR 3439 by @danielmewes, note that part of this still needs to be made configurable.

**Comment by VeXocide** [comment]
The configurable version is in CR 3448 by @danielmewes.

**Comment by danielmewes** [comment]
We're going to ship something different instead, that allows to configure a "waiting period" during which a server needs to hold its connection open, but doesn't fully join the cluster yet.

**Comment by danielmewes** [comment]
The join delay is in CR 3559

**Comment by danielmewes** [comment]
Merged into next b14e56f1dddb0b1df00ef190ac09f7b1eb96f7b2 . This introduces a `--join-delay <seconds>` option. I'll open a docs issue later (we can document this after shipping 2.3, it's not too critical).
